Fix XenServer Security Groups 'vmops' script

GabrielBrascher opened the pull request:
Fixing issue: #3192 Security Group rules not applied at all for
XenServer 6.5 / Advanced Zone

- Fix `tokens = line.split(':')` to `tokens = line.split(';')`
- Fix expected tokens size from 5 to 4
```
if len(tokens) != 5:
   continue
```
- Add logs
- remove unused vmops script. The XCP patch points to the vmops script on the parent folder [1]. Thus, all XenServer versions are considering the vmops script located at [2].

[1] https://github.com/apache/cloudstack/blob/master/scripts/vm/hypervisor/xenserver/xcposs/patch#L30
[2] https://github.com/apache/cloudstack/blob/master/scripts/vm/hypervisor/xenserver/vmops
## Description
<!--- Describe your changes in detail -->

<!-- For new features, provide link to FS, dev ML discussion etc. -->
<!-- In case of bug fix, the expected and actual behaviours, steps to reproduce. -->

<!-- When "Fixes: #<id>" is specified, the issue/PR will automatically be closed when this PR gets merged -->
<!-- For addressing multiple issues/PRs, use multiple "Fixes: #<id>" -->
<!-- Fixes: # -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):

## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

- SSH fail as expected
- Add Ingress and Egress rules with start port=22, end port=22, and cidr=0.0.0.0/0
- SSH succeeds as expected

**iptables -L**
```
Chain i-2-125-VM (1 references)
target     prot opt source               destination         
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:ssh ctstate NEW
DROP       all  --  anywhere             anywhere            

Chain i-2-125-VM-def (2 references)
target     prot opt source               destination         
RETURN     udp  --  anywhere             anywhere             PHYSDEV match --physdev-in vif22.0 --physdev-is-bridged match-set i-2-125-VM src udp dpt:domain
DROP       all  --  anywhere             anywhere             PHYSDEV match --physdev-in vif22.0 --physdev-is-bridged ! match-set i-2-125-VM src
DROP       all  --  anywhere             anywhere             PHYSDEV match --physdev-out vif22.0 --physdev-is-bridged ! match-set i-2-125-VM dst
i-2-125-VM-eg  all  --  anywhere             anywhere             PHYSDEV match --physdev-in vif22.0 --physdev-is-bridged match-set i-2-125-VM src
i-2-125-VM  all  --  anywhere             anywhere             PHYSDEV match --physdev-out vif22.0 --physdev-is-bridged

Chain i-2-125-VM-eg (1 references)
target     prot opt source               destination         
RETURN     tcp  --  anywhere             anywhere             tcp dpt:ssh ctstate NEW
DROP       all  --  anywhere             anywhere            

```
**cloud.log output**
```
DEBUG [root] Network rules compressed in base64 [eJzztCpJLrA2MgIhAz0w1DfQ8XONCFFwxS0FAG5GDOs=].
 INFO [root] Network rules to be processed [I:tcp;22;22;0.0.0.0/0,NEXT E:tcp;22;22;0.0.0.0/0,NEXT ].
 INFO [root] Programming network rules for vm  i-2-125-VM seqno=37 numrules=3 signature=5173d229b469dcfb63575ae93fde95ec guestIp=192.168.100.152, update iptables, reason=seqno_change_or_sig_change
DEBUG [root] Processing rule [I:tcp;22;22;0.0.0.0/0,NEXT].
DEBUG [root] Tokens ['I:tcp', '22', '22', '0.0.0.0/0,NEXT'].
 INFO [root] Ipset chaing [i-2-125-VM], VM [i-2-125-VM], Action [ACCEPT], Direction [src], egress rules [0]
DEBUG [root] port range [22:22]
DEBUG [root] ['iptables', '-I', 'i-2-125-VM', '-p', 'tcp', '-m', 'tcp', '--dport', '22:22', '-m', 'state', '--state', 'NEW', '-j', 'ACCEPT']
DEBUG [root] Processing rule [E:tcp;22;22;0.0.0.0/0,NEXT].
DEBUG [root] Tokens ['E:tcp', '22', '22', '0.0.0.0/0,NEXT'].
 INFO [root] Ipset chaing [i-2-125-VM-e], VM chain [i-2-125-VM-eg], VM name [i-2-125-VM], Action [RETURN], Direction [dst], egress rules [1]
DEBUG [root] port range [22:22]
DEBUG [root] ['iptables', '-I', 'i-2-125-VM-eg', '-p', 'tcp', '-m', 'tcp', '--dport', '22:22', '-m', 'state', '--state', 'NEW', '-j', 'RETURN']
DEBUG [root] Processing rule [].
DEBUG [root] Tokens [''].
WARNING [root] Network rule tokens size [1] is different from the expected size [4], ignoring rule ['']
DEBUG [root] Executing command: [['iptables', '-I', 'i-2-125-VM', '-p', 'tcp', '-m', 'tcp', '--dport', '22:22', '-m', 'state', '--state', 'NEW', '-j', 'ACCEPT']].
DEBUG [root] Executing command: [['iptables', '-I', 'i-2-125-VM-eg', '-p', 'tcp', '-m', 'tcp', '--dport', '22:22', '-m', 'state', '--state', 'NEW', '-j', 'RETURN']].
DEBUG [root] Writing log to /var/run/cloud/i-2-125-VM.log
```

<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document -->

GabrielBrascher commented:
Ping @andrijapanic @wido @rafaelweingartner @rhtyd @DaanHoogland @kiwiflyer @nathanejohnson @dhlaluku.

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-2615

wido commented on scripts/vm/hypervisor/xenserver/vmops:
```diff
@@ -1517,7 +1550,8 @@ def network_rules(session, args):
   except:
-    logging.debug("Failed to network rule !")
+    e = sys.exc_info()[0]
+    logging.error("Failed to network rule. Error: %s" % e)
```
Might want to use logging.exception here as you catch an Exception.

wido commented on scripts/vm/hypervisor/xenserver/vmops:
```diff
@@ -1438,14 +1448,31 @@ def network_rules(session, args):
+        end = tokens[2]
+        #cidrs = ['1.2.3.4/24', 'NEXT']
         cidrs = tokens.pop().split(",")
+        #cidrs = ['1.2.3.4/24']
```
All the above rules seem debugging lines, don't they?

andrijapanicsb commented:
Tested with XenServer 6.5sp1 - works fine, rules are added fine - all good.

LGTM

rafaelweingartner commented on scripts/vm/hypervisor/xenserver/vmops:
```diff
@@ -1438,14 +1448,31 @@ def network_rules(session, args):
+        #tokens[2] = '34'
+        end = tokens[2]
+        #cidrs = ['1.2.3.4/24', 'NEXT']
         cidrs = tokens.pop().split(",")
```
Why use pop? I mean, that is the same as `cidrs = tokens[3].split(",")`, which would follow the method you are using to access the array items.

rafaelweingartner commented on scripts/vm/hypervisor/xenserver/vmops:
```diff
@@ -1438,14 +1448,31 @@ def network_rules(session, args):
+        #cidrs = ['1.2.3.4/24', 'NEXT']
         cidrs = tokens.pop().split(",")
+        #cidrs = ['1.2.3.4/24']
         cidrs.pop()
```
This last pop needs some explanations. This is only here to remove the "NEXT " element in the array of CIDRs

GabrielBrascher commented:
@blueorangutan test

blueorangutan commented:
@GabrielBrascher a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

GabrielBrascher commented:
Thanks for the reviews @rafaelweingartner @wido @andrijapanicsb.
@wido and @rafaelweingartner I updated the code considering your comments.

blueorangutan commented:
<b>Trillian test result (tid-3411)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 33003 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3197-t3411-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_internal_lb.py
Intermittent failure detected: /marvin/tests/smoke/test_iso.py
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_templates.py
Intermittent failure detected: /marvin/tests/smoke/test_volumes.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermittent failure detected: /marvin/tests/smoke/test_host_maintenance.py
Smoke tests completed. 66 look OK, 4 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_04_extract_Iso | `Failure` | 1.06 | test_iso.py
test_04_extract_template | `Failure` | 1.06 | test_templates.py
test_06_download_detached_volume | `Failure` | 12.36 | test_volumes.py
test_04_rvpc_network_garbage_collector_nics | `Failure` | 271.34 | test_vpc_redundant.py

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-2616

GabrielBrascher commented:
@blueorangutan test

blueorangutan commented:
@GabrielBrascher a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-3412)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 33157 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3197-t3412-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_iso.py
Intermittent failure detected: /marvin/tests/smoke/test_templates.py
Intermittent failure detected: /marvin/tests/smoke/test_volumes.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 67 look OK, 3 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_04_extract_Iso | `Failure` | 1.10 | test_iso.py
test_04_extract_template | `Failure` | 1.11 | test_templates.py
test_06_download_detached_volume | `Failure` | 10.51 | test_volumes.py

GabrielBrascher commented:
All checks have passed, 2 LGTMs, manual tests passed and the Trillian tests that are failing are not relative to this change. I will be merging this PR.

This commit addresses the blocker #3192 raised for 4.12 RC3; thus, I will also continue with the VOTE process for 4.12 release candidate 4.

Thanks for the feedback guys!

andrijapanicsb commented:
Just FTR - in https://github.com/apache/cloudstack/issues/3192 there is another issue of SG rules not applied during VM live migration to new host - probably should be fixed here if possible.

rafaelweingartner commented on core/src/main/java/com/cloud/agent/api/SecurityGroupRulesCmd.java:
```diff
@@ -173,7 +173,7 @@ public String getSecIpsString() {
         final List<String> ips = getSecIps();
         if (ips == null) {
-            sb.append("0:");
+            sb.append("0;");
```
Shouldn't you use here `RULE_COMMAND_SEPARATOR`?

GabrielBrascher replied:
Yes, that would avoid future issues in case of an update on the RULE_COMMAND_SEPARATOR. I will update it soon when having the other raised points addressed as well. Thanks for pointing that!

rafaelweingartner commented on plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/wrapper/xenbase/CitrixMigrateCommandWrapper.java:
```diff
@@ -81,6 +81,13 @@ public Answer execute(final MigrateCommand command, final CitrixResourceBase cit
                 citrixResourceBase.migrateVM(conn, dsthost, vm, vmName);
                 vm.setAffinity(conn, dsthost);
+
+                final String result = citrixResourceBase.callHostPlugin(conn, "vmops", "destroy_network_rules_for_vm", "vmName", command.getVmName());
```
So, the cleanup was never implemented...?

GabrielBrascher replied:
Yes, it looks like @rafaelweingartner.

I am not aware if it was or not removing the network rules before 4.11/4.12, but I don't see how it would be removed as the mgmt was not calling `destroy_network_rules_for_vm` from the vmops.

rafaelweingartner replied:
Thanks for the explanation

rafaelweingartner commented on plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/wrapper/xenbase/CitrixMigrateCommandWrapper.java:
```diff
@@ -81,6 +81,13 @@ public Answer execute(final MigrateCommand command, final CitrixResourceBase cit
                 vm.setAffinity(conn, dsthost);
+
+                final String result = citrixResourceBase.callHostPlugin(conn, "vmops", "destroy_network_rules_for_vm", "vmName", command.getVmName());
+                if (result == null || result.isEmpty() || !Boolean.parseBoolean(result)) {
```
Have you tried using `BooleanUtils`?

GabrielBrascher replied:
Good point @rafaelweingartner. I will update to

```
if (BooleanUtils.toBoolean(result)) {
    s_logger.debug(String.format("Removed network rules from source host [%s] for migrated vm [%s]", dsthost.getHostname(conn), command.getVmName()));
} else {
    s_logger.warn(String.format("Failed to remove network rules from source host [%s] for migrated vm [%s]", dsthost.getHostname(conn), command.getVmName()));                
}
```

GabrielBrascher commented:
@wido @rafaelweingartner @andrijapanicsb and others. I have updated this PR addressing the issue reported on #3192.

Now the `CitrixMigrateCommandWrapper.java` calls the script `vmops` removing the SG network rules on the source host. Additionally, fixed a bug when secondary IP was null (was expecting `"0;"` instead of `"0:"`), adding the rules on the target host when migrating a VM.

andrijapanicsb commented:
thx @GabrielBrascher - will test today and let you know.

andrijapanicsb commented:
@blueorangutan package

blueorangutan commented:
@andrijapanicsb a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

GabrielBrascher commented:
Going to extract the added code to a method (executed if `citrixResourceBase.canBridgeFirewall()`), and add some JUnit tests.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-2619

GabrielBrascher commented:
@blueorangutan test

blueorangutan commented:
@GabrielBrascher a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

GabrielBrascher commented:
Updated PR with unit tests and addressing reviewers.

andrijapanicsb commented:
LGTM - adding rules work fine, also rules are migrated (and removed from source host) after VM live migration

^^^ on XenServer 6.5sp1

Thanks a lot @GabrielBrascher 

Any idea why/when things broke so badly ? (between 4.10 and 4.11) ?

GabrielBrascher commented:
@andrijapanic happy to be usefull :)

The problem is that we have some different workflows for XenServer and KVM. Thus, we struggled to cover all the possible cases.

The best approach, in my opinion, would be to have only one script handling SG network rules, regardless of the hypervisor. However, such change would require a deep refactoring on our codebase.

GabrielBrascher commented:
All checks (Jenkins + Travis) have passed and we have enough LGTMs.
I am on hold to check the CI results. If there are no objections I will be then merging this PR.

blueorangutan commented:
<b>Trillian test result (tid-3414)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 27293 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3197-t3414-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_internal_lb.py
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 68 look OK, 2 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_vpc_privategw_static_routes | `Failure` | 193.70 | test_privategw_acl.py
test_04_rvpc_network_garbage_collector_nics | `Failure` | 268.84 | test_vpc_redundant.py

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-2620

GabrielBrascher commented:
@blueorangutan test

blueorangutan commented:
@GabrielBrascher a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-3415)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 21617 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3197-t3415-kvm-centos7.zip
Smoke tests completed. 70 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
